subsys/bluetooth/audio/has.c

```diff
@@ -34,7 +56,362 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
+			       BT_GATT_PERM_READ_ENCRYPT, read_features, NULL, NULL),
 #if CONFIG_BT_HAS_PRESET_COUNT > 0
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_PRESET_CONTROL_POINT,
+#if defined(CONFIG_BT_EATT)
+			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY,
+#else
+			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,
+#endif /* CONFIG_BT_EATT */
+			       BT_GATT_PERM_WRITE_ENCRYPT, NULL, write_control_point, NULL),
+	BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_ACTIVE_PRESET_INDEX,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ_ENCRYPT,
+			       read_active_preset_index, NULL, NULL),
+	BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
+#endif /* CONFIG_BT_HAS_PRESET_COUNT > 0 */
+);
+
+#if CONFIG_BT_HAS_PRESET_COUNT > 0
+static struct has_client {
+	struct bt_conn *conn;
+	union {
+		struct bt_gatt_indicate_params ind;
+#if defined(CONFIG_BT_EATT)
+		struct bt_gatt_notify_params ntf;
+#endif /* CONFIG_BT_EATT */
+	} params;
+
+	struct bt_has_cp_read_presets_req read_presets_req;
+	struct k_work control_point_work;
+	struct k_work_sync control_point_work_sync;
+} has_client_list[BT_HAS_MAX_CONN];
+
+/* HAS internal preset representation */
+static struct bt_has_preset {
+	uint8_t index;
+	enum bt_has_properties properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+	char name[BT_HAS_PRESET_NAME_MAX + 1]; /* +1 byte for NULL-terminator */
+#else
+	const char *name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+} has_preset_list[CONFIG_BT_HAS_PRESET_COUNT];
+
+static void process_control_point_work(struct k_work *work);
+
+static struct has_client *client_get(struct bt_conn *conn)
+{
+	struct has_client *client = NULL;
+
+	for (size_t i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+
+		/* first free slot */
+		if (!client && !has_client_list[i].conn) {
+			client = &has_client_list[i];
+		}
+	}
+
+	__ASSERT(client, "failed to get client for conn %p", (void *)conn);
+
+	client->conn = bt_conn_ref(conn);
+
+	k_work_init(&client->control_point_work, process_control_point_work);
+
+	return client;
+}
+
+static bool read_presets_req_is_pending(struct has_client *client)
+{
+	return client->read_presets_req.num_presets > 0;
+}
+
+static void read_presets_req_free(struct has_client *client)
+{
+	client->read_presets_req.num_presets = 0;
+}
+
+static void client_free(struct has_client *client)
+{
+	(void)k_work_cancel(&client->control_point_work);
+
+	read_presets_req_free(client);
+
+	bt_conn_unref(client->conn);
+
+	client->conn = NULL;
+}
+
+static struct has_client *client_find(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+	}
+
+	return NULL;
+}
```

Comment:
This is very similar to `client_get`. Based on names only, it's tough to tell why we need both, which is perhaps another argument for a more explicit name for `client_get` to `client_get_or_new`, and then possible this can be renamed to `client_get`.